x-pack/plugins/xpack_main/server/lib/telemetry/monitoring/get_high_level_stats.js

```diff
@@ -187,8 +188,75 @@ export function getHighLevelStats(server, callCluster, clusterUuids, start, end,
  * @param {String} product The product to limit too ('kibana', 'logstash', 'beats')
  * @return {Promise} Response for the instances to fetch detailed for the product.
  */
-export function fetchHighLevelStats(server, callCluster, clusterUuids, start, end, product) {
+
+
+export function getLegacyParams(server, clusterUuids, start, end, product) {
   const config = server.config();
+
+  return {
+    index: getIndexPatternForStackProduct(product),
+    size: config.get('xpack.monitoring.max_bucket_size'),
+    ignoreUnavailable: true,
+    filterPath: [
+      'hits.hits._source.cluster_uuid',
+      `hits.hits._source.${product}_stats.${product}.version`,
+      `hits.hits._source.${product}_stats.os`,
+      `hits.hits._source.${product}_stats.usage`,
+      // we don't want metadata
+      `hits.hits._source.${product}_stats.cloud.name`,
+      `hits.hits._source.${product}_stats.cloud.id`,
+      `hits.hits._source.${product}_stats.cloud.vm_type`,
+      `hits.hits._source.${product}_stats.cloud.region`,
+      `hits.hits._source.${product}_stats.cloud.zone`
+    ],
+    body: {
+      query: createQuery({
+        start,
+        end,
+        type: `${product}_stats`,
+        filters: [
+          { terms: { cluster_uuid: clusterUuids } },
+          {
+            'bool': {
+              'must': [
+                { 'match': { 'kibana_stats.kibana.statuses.name': 'includes-usage-data' } },
+                { 'match': { 'kibana_stats.kibana.statuses.state': 'includes-usage-data' } }
+              ]
+            }
+          }
+        ]
+      }),
+      collapse: {
+        // a more ideal field would be the concatenation of the uuid + transport address for duped UUIDs (copied installations)
+        field: `${product}_stats.${product}.uuid`
+      },
+      sort: [
+        { 'timestamp': 'desc' }
+      ]
+    }
+  };
+}
+
+export function isLegacyVersion(version) {
```

Comment:
Can we get away without the version checks? I think there are 3 possible scenarios when fetching stats
- Monitoring cluster is new, was created with the new template.  We're good here
- Monitoring cluster is running 6.7 or 7.0 patch with the fix.  We're good here too with backwards compatibility changes
- Monitoring cluster and kibana are running old versions without any of these changes.  The old kibanas will still be updating those at a rate of every 10 seconds, and we don't need any filters on our query when fetching the data.  All the records will have usage data.  So I think we're good here to.

So I'm not sure I see the need for checking legacy versions? If we can avoid this we really should, it will make it very  hard to test and maintain.

Reply:
Maybe I'm not thinking through this right, but what about the scenario where we're running the new Kibana code (code in this PR with the `exists` query) but against monitoring indices with old mappings?

So, for instance, imagine a user is running Monitoring cluster 6.7.2 today, which has monitoring indices with the old mappings. Then they upgrade the Monitoring cluster and Kibana to 6.7.3 with this PR's code in it. The _template_ will get upgraded, affecting _new_ monitoring indices that get created going forward. However _existing_ monitoring indices' mappings won't be affected so the exists query would evaluate to `false`?

Reply:
@ycombinator yes I think we're thinking the same - new telemetry code + old indices without the mappings, in that case the query comes back empty so we end up with no usage telemetry data.  It sounds like you and @pickypg sorted that out with back-porting the mappings change - in that case the problem last 24 hours worst case, I was making it was a bigger problem than it really is.